static/js/robot.js

```diff
@@ -472,50 +399,78 @@ class Simulator{
 
     consistentLoop(period, func){
         /* Execute the robot at specificed frequency.
-
-        period (int): the period in seconds to run func in
+        period (int): the period in ms to run func in
         func (function): the function to execute each loop
-
         func may take only TIMEOUT_VALUE seconds to finish execution
         */
-        period = period * 1000;
         this.interval = setInterval(this.loopContent, period, func);
     }
 
     stop() {
-        if (this.isRunning == true) {
-            this.isRunning = false
+        if (this.mode !== "idle") {
+            this.mode = "idle"
             clearInterval(this.interval);
-            document.removeEventListener('keydown', down);
-            document.removeEventListener('keyup', up);
         }
+        postMessage({
+            mode: this.mode
+        })
     }
 
     simulateTeleop(){
         /* Simulate execution of the robot code.
-        Run setup_fn once before continuously looping loop_fn 
+        Run setup_fn once before continuously looping loop_fn
         TODO: Run teleop_setup once before looping teleop_main */
 
         this.robot = new RobotClass();
         this.loadStudentCode();
-        this.isRunning = true;
-
-        document.addEventListener('keydown', down);
-        document.addEventListener('keyup', up);
-        console.log("Simulate Teleop");
+        this.mode = "teleop"
         this.consistentLoop(this.robot.tickRate, this.teleop_main);
+        postMessage({
+            mode: this.mode
+        })
     }
 
     simulateAuto() {
         this.robot = new RobotClass();
         this.loadStudentCode();
-        this.isRunning = true;
-
-        // auto_thread = threading.Thread(group=null, target=this.autonomous_setup,
-        //                                 name="autonomous code thread", daemon=True)
-        // auto_thread.start()
+        this.mode = "auto"
         this.autonomous_setup()
-        this.consistentLoop(this.robot.tickRate, function(){});
-        setTimeout(function() { this.stop(); }, 30*1000);
+        setTimeout(function() { this.stop(); }.bind(this), 30*1000);
+        postMessage({
+            mode: this.mode
+        })
```

Comment:
Should we rely on the worker to send the mode to the main thread, or should we handle it independently in the main thread? Not sure